sparse/unit_test/Test_Sparse_extractCrsDiagonalBlocks.hpp

```diff
@@ -0,0 +1,156 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include "KokkosSparse_Utils.hpp"
+#include "KokkosKernels_TestUtils.hpp"
+
+namespace Test {
+template <typename scalar_t, typename lno_t, typename size_type,
+          typename device>
+void run_test_extract_diagonal_blocks(int nrows, int nblocks) {
+  using RowMapType     = Kokkos::View<size_type *, device>;
+  using EntriesType    = Kokkos::View<lno_t *, device>;
+  using ValuesType     = Kokkos::View<scalar_t *, device>;
+  using RowMapType_hm  = typename RowMapType::HostMirror;
+  using EntriesType_hm = typename EntriesType::HostMirror;
+  using ValuesType_hm  = typename ValuesType::HostMirror;
+  using crsMat_t       = CrsMatrix<scalar_t, lno_t, device, void, size_type>;
+
+  crsMat_t A;
+  std::vector<crsMat_t> DiagBlks(nblocks);
+
+  if (nrows != 0) {
+    // Generate test matrix
+    const size_type nnz = 2 + (nrows - 2) * 3 + 2;
+    RowMapType_hm hrow_map("hrow_map", nrows + 1);
+    EntriesType_hm hentries("hentries", nnz);
+    ValuesType_hm hvalues("hvalues", nnz);
+
+    // first row
+    hrow_map(0) = 0;
```

Comment:
I would suggest generating a random crs matrix each run. There are utilities for this in test utils.

Reply:
@e10harvey I just do a trick here by creating a crs matrix with ```values``` being the same as ```entries```. After extracting, in the diagonal blocks, the ```blk_entries``` change but the values in ```blk_values``` do not change. So, when checking the results, it will be easy to recover the ```blk_entries``` and check if these match with the ```blk_values```.

Reply:
Thanks, @vqd8a. Do you expect this function to work as intended with any CRS matrix who's column indices are in ascending order?

Reply:
@e10harvey Yes, it has to be a square CRS matrix with column indices in ascending order